src/main/java/cms/gov/madie/measure/services/MeasureService.java

```diff
@@ -589,4 +591,105 @@ private void updateOrganizationName(Measure measure) {
               });
     }
   }
+
+  public MeasureSet associateCmsId(String username, String qiCoreMeasureId, String qdmMeasureId) {
+    if (qiCoreMeasureId == null || qdmMeasureId == null) {
+      log.info(
+          "CMS ID could not be associated. Measure Ids [{}],[{}} cannot be null",
+          qiCoreMeasureId,
+          qdmMeasureId);
+      throw new InvalidIdException("CMS ID could not be associated. Please try again.");
+    }
+
+    Measure qiCoreMeasure = findMeasureById(qiCoreMeasureId);
+    Measure qdmMeasure = findMeasureById(qdmMeasureId);
+
+    if (qiCoreMeasure == null || qdmMeasure == null) {
+      log.info(
+          "CMS ID could not be associated. Measures with given Ids [{}],[{}] are not found",
+          qiCoreMeasureId,
+          qdmMeasureId);
+      throw new ResourceNotFoundException("CMS ID could not be associated. Please try again.");
+    }
+
+    validateCmsIdAssociation(username, qiCoreMeasure, qdmMeasure);
+
+    MeasureSet measureSet = measureSetService.findByMeasureSetId(qiCoreMeasure.getMeasureSetId());
+    measureSet.setCmsId(qdmMeasure.getMeasureSet().getCmsId());
+    measureSetRepository.save(measureSet);
+
+    return measureSet;
+  }
+
+  public List<Measure> checkDuplicateCmsId(Integer qdmCmsId) {
+    return measureRepository.findAllByModel(ModelType.QI_CORE.getValue()).stream()
```

Comment:
First, we are fetching all QICore measures and then iterating over all of them, and then fetching measureset for each of them. I think we can avoid multiple calls. this can be rewritten by using the aggregate/lookup pipeline. Please refer to  MeasureAclRepositoryImpl for example.

Reply:
Sure, I'll change that